Fixes #1286: String Collections support in $select.

anasik opened the pull request:
[EF Core 8 comes with support for primtive collections in models. ](https://learn.microsoft.com/en-us/ef/core/what-is-new/ef-core-8.0/whatsnew#primitive-collections)

Naturally, OData should pick up on that support. As of now, if you create a `List<string>` field in your model and then fetch all the records for that model, it works flawlessly. You see an array of strings in the response. 

However, if you try to `$select` that field, it fails with the following error: 

<details>
  <summary>Click to expand error</summary>
```
System.NullReferenceException: Object reference not set to an instance of an object.
   at Microsoft.EntityFrameworkCore.Query.RelationalShapedQueryCompilingExpressionVisitor.ShaperProcessingExpressionVisitor.CreateGetValueExpression(ParameterExpression dbDataReader, Int32 index, Boolean nullable, RelationalTypeMapping typeMapping, Type type, IPropertyBase property)
   at Microsoft.EntityFrameworkCore.Query.RelationalShapedQueryCompilingExpressionVisitor.ShaperProcessingExpressionVisitor.VisitExtension(Expression extensionExpression)
   at Microsoft.EntityFrameworkCore.Query.RelationalShapedQueryCompilingExpressionVisitor.ShaperProcessingExpressionVisitor.ProcessShaper(Expression shaperExpression, RelationalCommandCache& relationalCommandCache, IReadOnlyList`1& readerColumns, LambdaExpression& relatedDataLoaders, Int32& collectionId)
   at Microsoft.EntityFrameworkCore.Query.RelationalShapedQueryCompilingExpressionVisitor.ShaperProcessingExpressionVisitor.VisitExtension(Expression extensionExpression)
   at System.Linq.Expressions.ExpressionVisitor.VisitMemberAssignment(MemberAssignment node)
   at System.Linq.Expressions.ExpressionVisitor.VisitMemberBinding(MemberBinding node)
   at System.Linq.Expressions.ExpressionVisitor.Visit[T](ReadOnlyCollection`1 nodes, Func`2 elementVisitor)
   at System.Linq.Expressions.ExpressionVisitor.VisitMemberInit(MemberInitExpression node)
   at System.Linq.Expressions.ExpressionVisitor.VisitMemberAssignment(MemberAssignment node)
   at System.Linq.Expressions.ExpressionVisitor.VisitMemberBinding(MemberBinding node)
   at System.Linq.Expressions.ExpressionVisitor.Visit[T](ReadOnlyCollection`1 nodes, Func`2 elementVisitor)
   at System.Linq.Expressions.ExpressionVisitor.VisitMemberInit(MemberInitExpression node)
   at Microsoft.EntityFrameworkCore.Query.RelationalShapedQueryCompilingExpressionVisitor.ShaperProcessingExpressionVisitor.ProcessShaper(Expression shaperExpression, RelationalCommandCache& relationalCommandCache, IReadOnlyList`1& readerColumns, LambdaExpression& relatedDataLoaders, Int32& collectionId)
   at Microsoft.EntityFrameworkCore.Query.RelationalShapedQueryCompilingExpressionVisitor.VisitShapedQuery(ShapedQueryExpression shapedQueryExpression)
   at Microsoft.EntityFrameworkCore.Query.ShapedQueryCompilingExpressionVisitor.VisitExtension(Expression extensionExpression)
   at Microsoft.EntityFrameworkCore.Query.RelationalShapedQueryCompilingExpressionVisitor.VisitExtension(Expression extensionExpression)
   at Microsoft.EntityFrameworkCore.Query.QueryCompilationContext.CreateQueryExecutor[TResult](Expression query)
   at Microsoft.EntityFrameworkCore.Storage.Database.CompileQuery[TResult](Expression query, Boolean async)
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.CompileQueryCore[TResult](IDatabase database, Expression query, IModel model, Boolean async)
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.<>c__DisplayClass12_0`1.<ExecuteAsync>b__0()
   at Microsoft.EntityFrameworkCore.Query.Internal.CompiledQueryCache.GetOrAddQuery[TResult](Object cacheKey, Func`1 compiler)
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.ExecuteAsync[TResult](Expression query, CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.Query.Internal.EntityQueryProvider.ExecuteAsync[TResult](Expression expression, CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.Query.Internal.EntityQueryable`1.GetAsyncEnumerator(CancellationToken cancellationToken)
   at System.Text.Json.Serialization.Converters.IAsyncEnumerableOfTConverter`2.OnWriteResume(Utf8JsonWriter writer, TAsyncEnumerable value, JsonSerializerOptions options, WriteStack& state)
   at System.Text.Json.Serialization.JsonCollectionConverter`2.OnTryWrite(Utf8JsonWriter writer, TCollection value, JsonSerializerOptions options, WriteStack& state)
   at System.Text.Json.Serialization.JsonConverter`1.TryWrite(Utf8JsonWriter writer, T& value, JsonSerializerOptions options, WriteStack& state)
   at System.Text.Json.Serialization.JsonConverter`1.WriteCore(Utf8JsonWriter writer, T& value, JsonSerializerOptions options, WriteStack& state)
   at System.Text.Json.Serialization.Metadata.JsonTypeInfo`1.SerializeAsync(Stream utf8Json, T rootValue, CancellationToken cancellationToken, Object rootValueBoxed)
   at System.Text.Json.Serialization.Metadata.JsonTypeInfo`1.SerializeAsync(Stream utf8Json, T rootValue, CancellationToken cancellationToken, Object rootValueBoxed)
   at System.Text.Json.Serialization.Metadata.JsonTypeInfo`1.SerializeAsync(Stream utf8Json, T rootValue, CancellationToken cancellationToken, Object rootValueBoxed)
   at Microsoft.AspNetCore.Mvc.Formatters.SystemTextJsonOutputFormatter.WriteResponseBodyAsync(OutputFormatterWriteContext context, Encoding selectedEncoding)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeNextResultFilterAsync>g__Awaited|30_0[TFilter,TFilterAsync](ResourceInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.Rethrow(ResultExecutedContextSealed context)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.ResultNext[TFilter,TFilterAsync](State& next, Scope& scope, Object& state, Boolean& isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.InvokeResultFilters()
--- End of stack trace from previous location ---
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeFilterPipelineAsync>g__Awaited|20_0(ResourceInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeAsync>g__Awaited|17_0(ResourceInvoker invoker, Task task, IDisposable scope)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeAsync>g__Awaited|17_0(ResourceInvoker invoker, Task task, IDisposable scope)
   at Microsoft.AspNetCore.Authorization.AuthorizationMiddleware.Invoke(HttpContext context)
   at Swashbuckle.AspNetCore.SwaggerUI.SwaggerUIMiddleware.Invoke(HttpContext httpContext)
   at Swashbuckle.AspNetCore.Swagger.SwaggerMiddleware.Invoke(HttpContext httpContext, ISwaggerProvider swaggerProvider)
   at Microsoft.AspNetCore.Authentication.AuthenticationMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.Diagnostics.DeveloperExceptionPageMiddlewareImpl.Invoke(HttpContext context)
```
</details>

This happens because in `ProjectAsWrapper`, we check if it's a collection, and if yes, we wrap it as a collection rather than an element. 
- With SQL Server this works perfectly fine for all collections except for strings.
- With SQLite, pretty much all primitive collections fail.

That's because queries of the form: `Entity().Select( d=> d.StringListProperty.Select(d=> d)))` just inherently don't work. It appears to be a bug in EF Core that one can easily recreate without even needing to initialize OData. 

However, for our purposes, we can easily get around this by simply wrapping the List as an element as opposed to a collection.

I tested this as much as I could and so far I haven't discovered any negative side-effects of this. It's working perfectly fine and I am even able to filter on the string list using any/all .

Fixes #1286

anasik commented:
@microsoft-github-policy-service agree

xuzhg commented:
Do you mind to add tests

anasik commented:
> Do you mind to add tests

I can. I'm just not sure where. Should I create a folder named "Strings" under E2E?

anasik commented:
@xuzhg Since primitive collections only started getting supported as recent as EF Core 8, I'm gonna have to add net8 as a target framework and add a test that only runs on net8.

xuzhg commented:
> I'm just not sure where. Should I create a folder named "Strings" unde

We have '[Microsoft.AspNetCore.OData.Tests](https://github.com/OData/AspNetCoreOData/tree/main/test/Microsoft.AspNetCore.OData.Tests)' for Unit test, and 

[Microsoft.AspNetCore.OData.E2E.Tests](https://github.com/OData/AspNetCoreOData/tree/main/test/Microsoft.AspNetCore.OData.E2E.Tests) for End to End test.

Can you find a folder to add the tests within the E2E test?

xuzhg commented:
> @xuzhg Since primitive collections only started getting supported as recent as EF Core 8, I'm gonna have to add net8 as a target framework and add a test that only runs on net8. Is that okay?

I am thinking it. Add '.NET8' target framework to test project will run all the test cases on this target framework. Basically, it's ok. 
However, We'd make sure the CI/CD can handle it. You can try that and let's review it.

By the way, can we mock the 'EF Core 8' primitive collection? Or can we add a sample project to play the EF Core 8?

habbes commented:
@xuzhg AspNetCore OData 9 preview uses .NET 8 as the target framework

anasik commented:
> > @xuzhg Since primitive collections only started getting supported as recent as EF Core 8, I'm gonna have to add net8 as a target framework and add a test that only runs on net8. Is that okay?
> 
> I am thinking it. Add '.NET8' target framework to test project will run all the test cases on this target framework. Basically, it's ok. However, We'd make sure the CI/CD can handle it. You can try that and let's review it.
> 
> By the way, can we mock the 'EF Core 8' primitive collection? Or can we add a sample project to play the EF Core 8?

anasik commented:
> I am thinking it. Add '.NET8' target framework to test project will run all the test cases on this target framework. Basically, it's ok. However, We'd make sure the CI/CD can handle it. You can try that and let's review it.

@xuzhg, Yeah I have already added the NET8 target framework locally, I just wanted to ask first. 

> By the way, can we mock the 'EF Core 8' primitive collection? Or can we add a sample project to play the EF Core 8?

I'm not quite sure what you mean by this but I did add a model for testing purposes that has all primitive collection fields.

> Can you find a folder to add the tests within the E2E test?

I decided to just create a new one specifically for Lists. I explored a lot of existing folders and none seemed like a good enough candidate for such a specific case.

anasik commented:
@xuzhg, 

I have added the test. Assuming you have net8.0 installed on your machine, you can directly run said test through the following command inside the `E2E` folder:

`dotnet test --filter DisplayName~ListsTest -f net8.0`

A bunch of observations you could make here:
- I added SQLite to the project packages and a *.db file. That's a SQLite DB. I had to include it because the bug in question is limited to SQL. I originally encountered it on an SQL Server instance but SQLite seemed much more friendly for automated testing.
- I also added some more cases to the  `SelectExpandBinder` file. That's because when working with SQLite, I noticed that, unlike SQL Server, it wasn't just `List<string>` that was failing but also every other primitive collection. 
- For some of the test theories I created, I left the `List<Uri>` case commented out because it's causing the tests to fail. Even in the original description I added to these PRs, I hade made some observations about some undefined behaviour around the Uri type and like the other problems, it's just a lot more pronounced in SQLite. Sadly, it felt a little too outside the scope of this branch for me to continue investigating just yet.

habbes commented on src/Microsoft.AspNetCore.OData/Query/Expressions/SelectExpandBinder.cs:
```diff
@@ -100,13 +100,14 @@ public virtual Expression BindSelectExpand(SelectExpandClause selectExpandClause
                 if (elementType == typeof(string)
                     || elementType == typeof(Uri)
                     || elementType == typeof(DateTime)
+#if NET6_0_OR_GREATER
```
In my view, this is a good indicator that we should consider moving this feature to AspNetCoreOData 9.

anasik replied:
I am currently architecting a large-scale project that heavily uses primitive collections. It was in the middle of work that I both discovered and fixed this. As of now, we are using a modified package reference in our repository to get around this. If we fix it in AspNetCoreOData 8, we will no longer have to do that. 

We can obviously continue to do so but I'm just highlighting that this is an issue that affects production _and_ it's already been addressed sufficiently here. For a week or so, I have virtually only been heavily testing this to ensure there's no negative effects and I plan to do a little more testing on my own to be even more sure.

There could be other projects like mine out there. Some of them closer to completion than others where upgrading from net8 to net9 may not be feasible right now. 

That being said, you'd have seen that initially I made the whole block conditional for only net8 but then I had to remove that condition when I realize d that the project doesn't target net8 and therefore that block would be ignored during compilation.  

But I think we can also conclude that the block wouldn't affect anything below net8 since it only affects primitive collections which aren't supported below net8 anyway.

habbes commented on test/Microsoft.AspNetCore.OData.E2E.Tests/Lists/ListsTest.cs:
```diff
@@ -253,8 +253,8 @@ public async Task QueryEntitySelect(string format, string select)
         [InlineData("application/json;odata.metadata=full", "ListTestInt", "/all(i: i le 5)", 1)]
-        [InlineData("application/json;odata.metadata=full", "ListTestDateTime", "/any(d: d gt 2024-07-26T00:00:00Z)", 4)]
-        [InlineData("application/json;odata.metadata=full", "ListTestDateTime", "/all(d: d gt 2024-07-26T00:00:00Z)", 4)]
+        //[InlineData("application/json;odata.metadata=full", "ListTestDateTime", "/any(d: d gt 2024-07-26T00:00:00Z)", 4)]
```
Why are these commented out?

anasik replied:
> For some of the test theories I created, I left the `List<Uri>` case commented out because it's causing the tests to fail. Even in the original description I added to these PRs, I hade made some observations about some undefined behavior around the `Uri ` type and like the other problems, it's just a lot more pronounced in SQLite. Sadly, it felt a little too outside the scope of this branch for me to continue investigating just yet.
Similarly, for the `$filter` theory, I also commented out the `List<DateTimeOffset>` tests. It doesn't seem like they can be passed since the exception thrown insists on using client evaluation.

I explained this here in the last comment. But basically:

- Both `DateTimeOffset` and `Uri` are a little more complicated than the rest. I've been experiencing a lot of weird behavior with `List<Uri>` even in the absence of OData. I've been meaning to investigate this a bit more but I feel like it lies outside the scope of this PR.
- If you run the commented out `List<DateTimeOffset>` tests, you'd see a very verbose exception which instructs you to convert the `DbSet` to `IEnumerable` or `IList` before querying, which basically means that it's both not exactly a bug as much as an unimplemented feature and also outside the scope of this PR.
- To add a little more context, deferred execution is a prerequisite for this PR because the bug in question cannot be recreated with in-memory lists. It happens only when we attempt to run certain queries directly on the `DbSet`. Which is why an exception telling me to convert the `DbSet` to an `Enumerable` is effectively telling me to move on. 

Since I plan to at least attempt to address both of these issues in a subsequent PR, I chose to comment these tests out instead of straight up deleting them so that I can just pass them later. 

I think a couple of them might actually pass if the test was using SQL Server instead of SQLite, but then again, half of these cases won't even be needed on SQL Server. Furthermore, even on SQL Server, `List<Uri>` behaves weirdly. Except it just doesn't crash.  Just gives wrong/bad output.

habbes replied:
I think the context could be added to the comments so that it's clear from reading the code why the code is commented and what the pending work is to address them.

WanjohiSammy commented on test/Microsoft.AspNetCore.OData.E2E.Tests/Lists/ListsDataModel.cs:
```diff
@@ -0,0 +1,30 @@
+        public IList<DateTimeOffset> ListTestDateTime { get; set; }
+        public IList<Uri> ListTestUri { get; set; }
+        public uint[] ListTestUint { get; set; }
+    }
```
Can you kindly add ListOfOders property `IList<Order>? ListTestOrder` and make it nullable.
```suggestion
       public IList<Order>? ListTestOrders { get; set; }
    }
    public class Order
    {
        [Key]
        public string OrderId { get; set; }
    }
```

WanjohiSammy commented on test/Microsoft.AspNetCore.OData.E2E.Tests/Lists/ListsTest.cs:
```diff
@@ -0,0 +1,289 @@
+        [InlineData("application/json;odata.metadata=full", "ListTestDouble")]
+        [InlineData("application/json;odata.metadata=full", "ListTestInt")]
+        [InlineData("application/json;odata.metadata=full", "ListTestDateTime")]
+        //[InlineData("application/json;odata.metadata=full", "ListTestUri")]
```
Once you have the Property `ListTestOrders `Add this line:
`[InlineData("application/json;odata.metadata=full", "ListTestOrders")]`

WanjohiSammy commented on test/Microsoft.AspNetCore.OData.E2E.Tests/Lists/ListsTest.cs:
```diff
@@ -0,0 +1,289 @@
+        [InlineData("application/json;odata.metadata=full", "ListTestDouble")]
+        [InlineData("application/json;odata.metadata=full", "ListTestInt")]
+        [InlineData("application/json;odata.metadata=full", "ListTestDateTime")]
+        //[InlineData("application/json;odata.metadata=full", "ListTestUri")]
```
Same here: Add this line:
`[InlineData("application/json;odata.metadata=full", "ListTestOrders")]`

WanjohiSammy commented on test/Microsoft.AspNetCore.OData.E2E.Tests/Lists/ListsController.cs:
```diff
@@ -0,0 +1,179 @@
+                   ListTestFloat = new List<float> { 1.1f, 2.2f },
+                   ListTestDateTime = new List<DateTimeOffset> { DateTime.Now, DateTime.UtcNow },
+                   ListTestUri = new List<Uri> { new Uri("https://example.com") },
+                   ListTestUint = new uint[] { 1, 2, 3 }
```
Initialize `ListTestOrders` with data here, and leave `ListTestOrders` as null for the other products 
```suggestion
                   ListTestUint = new uint[] { 1, 2, 3 },
                   ListTestOrders = new List<Order>
                    {
                        new Order { OrderId = "Order1" },
                        new Order { OrderId = "Order2" }
                    }
```

WanjohiSammy commented on test/Microsoft.AspNetCore.OData.E2E.Tests/Microsoft.AspNetCore.OData.E2E.Tests.csproj:
```diff
@@ -18,17 +18,18 @@
-    <PackageReference Include="Moq" Version="4.13.1" />
-    <PackageReference Include="xunit" Version="2.3.1" />
-    <PackageReference Include="xunit.runner.visualstudio" Version="2.3.1" />
+    <PackageReference Include=" Microsoft.AspNetCore.Mvc.Testing" Version=" 3.1.7" />
```
Will this extra space not be bringing other issues?

anasik replied:
I honestly have no idea what this is. I can't even see this within my editor. If there's a way to undo it , I will. But no it doesn't cause problems.

anasik replied:
Fixed it. Previously I was using Visual Studio Mac, and that's what added these spaces and wouldn't allow me to remove them. Same for VS Code. I just managed to remove them easily using Rider.

WanjohiSammy commented on src/Microsoft.AspNetCore.OData/Query/Expressions/SelectExpandBinder.cs:
```diff
@@ -97,7 +97,36 @@ public virtual Expression BindSelectExpand(SelectExpandClause selectExpandClause
+                    || elementType == typeof(ulong)
+                    || elementType == typeof(short)
+                    || elementType == typeof(ushort)
+                    || elementType == typeof(Guid))
```
Move these entire conditions to TypeHelper method and encapsulate in the method:

```cs
public static bool IsPrimitiveOrKnownType(Type type)
{
    return type.IsPrimitive
        || type == typeof(string)
#if NET6_0_OR_GREATER
        || type == typeof(TimeOnly)
        || type == typeof(DateOnly)
#endif
        || type == typeof(Uri)
        || type == typeof(DateTime)
        || type == typeof(DateTimeOffset)
        || type == typeof(Decimal)
        || type == typeof(Guid);
}
```

WanjohiSammy commented on src/Microsoft.AspNetCore.OData/Query/Expressions/SelectExpandBinder.cs:
```diff
@@ -97,7 +97,36 @@ public virtual Expression BindSelectExpand(SelectExpandClause selectExpandClause
+
+            if (isCollection)
+            {
+                isPrimitiveCollection = TypeHelper.IsPrimitive(elementType);
```
Should be:
```suggestion
                isPrimitiveCollection = TypeHelper.IsPrimitiveOrKnownType(elementType);
```

Move these checks to `TypeHelper `inside `IsPrimitiveOrKnownType()` method
```cs
if(elementType == typeof(Uri)
                    || elementType == typeof(DateTime)
#if NET6_0_OR_GREATER
 || elementType == typeof(DateOnly)
#endif
                    || elementType == typeof(DateTimeOffset)
                    || elementType == typeof(int)
                    || elementType == typeof(uint)
                    || elementType == typeof(long)
                    || elementType == typeof(bool)
                    || elementType == typeof(double)
                    || elementType == typeof(float)
                    || elementType == typeof(byte)
                    || elementType == typeof(sbyte)
                    || elementType == typeof(char)
                    || elementType == typeof(decimal)
                    || elementType == typeof(ulong)
                    || elementType == typeof(short)
                    || elementType == typeof(ushort)
                    || elementType == typeof(Guid))
```

The `IsPrimitiveOrKnownType()` method in TypeHelper method can look like this:

```c#
/// <summary>
/// Check whether the given type is a primitive type or known type.
/// </summary>
/// <param name="type">The type to validate.</param>
// <returns>True if type primitive or known types, otherwise False.</returns>
public static bool IsPrimitiveOrKnownType(Type type)
{
    return type.IsPrimitive
        || type == typeof(string)
#if NET6_0_OR_GREATER
        || type == typeof(TimeOnly)
        || type == typeof(DateOnly)
#endif
        || type == typeof(Uri)
        || type == typeof(DateTime)
        || type == typeof(DateTimeOffset)
        || type == typeof(Decimal)
        || type == typeof(Guid);
}
```

anasik replied:
That's exactly what it looks like.

WanjohiSammy commented on test/Microsoft.AspNetCore.OData.E2E.Tests/Lists/ListsDataModel.cs:
```diff
@@ -0,0 +1,36 @@
+        public uint[] ListTestUint { get; set; }
+       public IList<Order>? ListTestOrders { get; set; }
+    }
+    public class Order
```
Add this table to SQLite DB as well

anasik replied:
Already did. Along with a lot of other things that may have escaped your mind. I only pushed my changes when all tests passed.

anasik commented:
Already have

On Fri, Aug 9, 2024, 18:35 Samuel Wanjohi ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In test/Microsoft.AspNetCore.OData.E2E.Tests/Lists/ListsDataModel.cs
> <https://github.com/OData/AspNetCoreOData/pull/1282#discussion_r1711315790>
> :
>
> > @@ -26,5 +26,11 @@ public class Product
>          public IList<DateTimeOffset> ListTestDateTime { get; set; }
>          public IList<Uri> ListTestUri { get; set; }
>          public uint[] ListTestUint { get; set; }
> +       public IList<Order>? ListTestOrders { get; set; }
>
> nit: Indentation
> ------------------------------
>
> In src/Microsoft.AspNetCore.OData/Query/Expressions/SelectExpandBinder.cs
> <https://github.com/OData/AspNetCoreOData/pull/1282#discussion_r1711460914>
> :
>
> > +#endif
> +                    || elementType == typeof(DateTimeOffset)
> +                    || elementType == typeof(int)
> +                    || elementType == typeof(uint)
> +                    || elementType == typeof(long)
> +                    || elementType == typeof(bool)
> +                    || elementType == typeof(double)
> +                    || elementType == typeof(float)
> +                    || elementType == typeof(byte)
> +                    || elementType == typeof(sbyte)
> +                    || elementType == typeof(char)
> +                    || elementType == typeof(decimal)
> +                    || elementType == typeof(ulong)
> +                    || elementType == typeof(short)
> +                    || elementType == typeof(ushort)
> +                    || elementType == typeof(Guid))
>
> Move these entire conditions to TypeHelper method and encapsulate in the
> method:
>
> public static bool IsPrimitiveOrKnownType(Type type){
>     return type.IsPrimitive
>         || type == typeof(string)
> #if NET6_0_OR_GREATER
>         || type == typeof(TimeOnly)
>         || type == typeof(DateOnly)
> #endif
>         || type == typeof(Uri)
>         || type == typeof(DateTime)
>         || type == typeof(DateTimeOffset)
>         || type == typeof(Decimal)
>         || type == typeof(Guid);}
>
> ------------------------------
>
> In src/Microsoft.AspNetCore.OData/Query/Expressions/SelectExpandBinder.cs
> <https://github.com/OData/AspNetCoreOData/pull/1282#discussion_r1711476106>
> :
>
> > @@ -97,7 +97,36 @@ public virtual Expression BindSelectExpand(SelectExpandClause selectExpandClause
>              int? modelBoundPageSize = null)
>          {
>              Type elementType;
> -            bool isCollection = TypeHelper.IsCollection(source.Type, out elementType);
> +            bool isCollection, isPrimitiveCollection = false;
> +            isCollection = TypeHelper.IsCollection(source.Type, out elementType);
> +
> +            if (isCollection)
> +            {
> +                isPrimitiveCollection = TypeHelper.IsPrimitive(elementType);
>
> Should be:
> ⬇️ Suggested change
>
> -                isPrimitiveCollection = TypeHelper.IsPrimitive(elementType);
> +                isPrimitiveCollection = TypeHelper.IsPrimitiveOrKnownType(elementType);
>
> Move these checks to TypeHelper inside IsPrimitiveOrKnownType() method
>
> if(elementType == typeof(Uri)
>                     || elementType == typeof(DateTime)
> #if NET6_0_OR_GREATER
>  || elementType == typeof(DateOnly)
> #endif
>                     || elementType == typeof(DateTimeOffset)
>                     || elementType == typeof(int)
>                     || elementType == typeof(uint)
>                     || elementType == typeof(long)
>                     || elementType == typeof(bool)
>                     || elementType == typeof(double)
>                     || elementType == typeof(float)
>                     || elementType == typeof(byte)
>                     || elementType == typeof(sbyte)
>                     || elementType == typeof(char)
>                     || elementType == typeof(decimal)
>                     || elementType == typeof(ulong)
>                     || elementType == typeof(short)
>                     || elementType == typeof(ushort)
>                     || elementType == typeof(Guid))
>
> The IsPrimitiveOrKnownType() method in TypeHelper method can look like
> this:
>
> /// <summary>/// Check whether the given type is a primitive type or known type./// </summary>/// <param name="type">The type to validate.</param>// <returns>True if type primitive or known types, otherwise False.</returns>public static bool IsPrimitiveOrKnownType(Type type){
>     return type.IsPrimitive
>         || type == typeof(string)
> #if NET6_0_OR_GREATER
>         || type == typeof(TimeOnly)
>         || type == typeof(DateOnly)
> #endif
>         || type == typeof(Uri)
>         || type == typeof(DateTime)
>         || type == typeof(DateTimeOffset)
>         || type == typeof(Decimal)
>         || type == typeof(Guid);}
>
> ------------------------------
>
> In test/Microsoft.AspNetCore.OData.E2E.Tests/Lists/ListsDataModel.cs
> <https://github.com/OData/AspNetCoreOData/pull/1282#discussion_r1711478316>
> :
>
> > +    {
> +        [Key]
> +        public string ProductId { get; set; }
> +        public string Name { get; set; }
> +        public string Category { get; set; }
> +        public IList<string> ListTestString { get; set; } = new List<string>();
> +        public IList<bool> ListTestBool { get; set; }
> +        public IList<int> ListTestInt { get; set; }
> +        public IList<double> ListTestDouble { get; set; }
> +        public IList<float> ListTestFloat { get; set; }
> +        public IList<DateTimeOffset> ListTestDateTime { get; set; }
> +        public IList<Uri> ListTestUri { get; set; }
> +        public uint[] ListTestUint { get; set; }
> +       public IList<Order>? ListTestOrders { get; set; }
> +    }
> +    public class Order
>
> Add this table to SQLite DB as well
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/OData/AspNetCoreOData/pull/1282#pullrequestreview-2229996086>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAWOXU4WN6JIVSC67AJFAR3ZQTATLAVCNFSM6AAAAABK6WMKRWVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDEMRZHE4TMMBYGY>
> .
> You are receiving this because you modified the open/close state.Message
> ID: ***@***.***>
>

anasik commented on test/Microsoft.AspNetCore.OData.E2E.Tests/Lists/ListsDataModel.cs:
```diff
@@ -26,5 +26,11 @@ public class Product
         public IList<DateTimeOffset> ListTestDateTime { get; set; }
         public IList<Uri> ListTestUri { get; set; }
         public uint[] ListTestUint { get; set; }
+       public IList<Order>? ListTestOrders { get; set; }
```
Yeah I fixed it right after I committed your suggestion.

WanjohiSammy commented on src/Microsoft.AspNetCore.OData/Common/TypeHelper.cs:
```diff
@@ -252,6 +252,26 @@ Type collectionInterface
+        /// <summary>
+        /// Checks if element is Primitive
+        /// </summary>
+        /// <param name="elementType"></param>
```
Complete on the doc. What is the param, what is being returned?
```suggestion
        /// <summary>
        /// Check whether the given type is a primitive type or known type.
        /// </summary>
        /// <param name="type">The type to validate.</param>
        /// <returns>True if type is primitive or known type, otherwise False.</returns>
```

WanjohiSammy commented on test/Microsoft.AspNetCore.OData.E2E.Tests/Lists/ListsTest.cs:
```diff
@@ -0,0 +1,292 @@
+        {
+            string currentDirectory = Environment.CurrentDirectory;
+            //Console.WriteLine("Current Working Directory: " + currentDirectory);
+            //throw new InvalidOperationException("Current Working Directory: " + currentDirectory);
```
You can remove these commented lines if they are not in used

WanjohiSammy commented on test/Microsoft.AspNetCore.OData.E2E.Tests/Lists/ListsContext.cs:
```diff
@@ -0,0 +1,19 @@
+﻿using System;
+using System.IO;
+using Microsoft.EntityFrameworkCore;
+
```
You forgot to add Microsoft copyright comment here. You can refer the other files.

WanjohiSammy commented:
LGTM

WanjohiSammy commented on test/Microsoft.AspNetCore.OData.E2E.Tests/Lists/ListsTest.cs:
```diff
@@ -0,0 +1,290 @@
+            string currentDirectory = Environment.CurrentDirectory;
+            services.ConfigureControllers(typeof(ProductsController), typeof(MetadataController));
+            services.AddDbContext<ListsContext>(options =>
+            options.UseSqlite("Data Source=../../../../test/Microsoft.AspNetCore.OData.E2E.Tests/Lists/DataBaseSrc.db"));
```
Is it possible to use in-memory sqlite instead of check-in binary .db file to the source code. Have you tried this https://learn.microsoft.com/en-us/dotnet/standard/data/sqlite/in-memory-databases

anasik replied:
@WanjohiSammy, 
Okay, I tried it and I have a working example with `Data Source=:memory:`. I had to add another few lines to make it work. But I haven't pushed it yet because I'm waiting on @ElizabethOkerio and @habbes response.

WanjohiSammy replied:
@anasik If you have a way of not using the in-file db, please use it. We do not allow the binary .db file.

anasik replied:
As you wish.

WanjohiSammy commented on .gitattributes:
```diff
@@ -11,6 +11,7 @@
 *.sql binary
 *.nupkg binary
 *.exe binary
+*.db binary
```
You can also remove this.

anasik replied:
Sure. But we could also leave this here in case someone needs to upload a db in the future and cannot for the life of them figure out why it gets corrupted with every update? This change won't affect anything for now but could offer convenience in the future to someone.

anasik commented:
@xuzhg, I just wanted to ask, are you guys waiting on me for something on this?

anasik commented on src/Microsoft.AspNetCore.OData/Common/TypeHelper.cs:
```diff
@@ -252,6 +252,26 @@ Type collectionInterface
+        /// </summary>
+        /// <param name="type">The type to validate.</param>
+        /// <returns>True if type is primitive or known type, otherwise False.</returns>
+        public static bool IsPrimitiveOrKnownType(Type type)
```
@WanjohiSammy, I am well-aware of the correct primitive types but I had to cover all the types that are not only affected by the bug but were also documented to support Primitive Collections in the EF Core 8 change log.

That being said, I took a look at said method it's not immediately evident to me exactly which types of mine you're concerned about. 

Since you've already mentioned Uri, 
- [Here](https://learn.microsoft.com/en-us/ef/core/what-is-new/ef-core-8.0/whatsnew#primitive-collection-properties) it is being given express treatment therefore I had to support it.
- It's also affected by the bug at hand. 
- In fact, this type is actually the most buggy of all, as  I have mentioned several times in earlier interactions. In fact, I was hoping this would get merged quickly so I could debug Uri more.

habbes commented on test/Microsoft.AspNetCore.OData.E2E.Tests/Lists/ListsEdmModel.cs:
```diff
@@ -0,0 +1,29 @@
+            return edmModel;
+        }
+
+           }
```
Nitpick: incorrect indentation

WanjohiSammy commented on src/Microsoft.AspNetCore.OData/Common/TypeHelper.cs:
```diff
@@ -252,6 +252,26 @@ Type collectionInterface
+        public static bool IsPrimitiveOrKnownType(Type type)
+        {
+            return type.IsPrimitive 
+                   || type == typeof(string)
```
Also note a change in `Microsoft.AspNetCore.OData.xml` when you build the application. Please check-in

```xml
 <member name="M:Microsoft.AspNetCore.OData.Common.TypeHelper.IsPrimitiveOrKnownType(System.Type)">
     <summary>
     Check whether the given type is a primitive type or known type.
     </summary>
     <param name="type">The type to validate.</param>
     <returns>True if type is primitive or known type, otherwise False.</returns>
 </member>
```

anasik replied:
I'm not sure if this comment was for me but I rebuilt the solution to see what it looks like and it's exactly like the snippet you posted.

I'm a little confused about what I need to do here?

WanjohiSammy replied:
Push the change in that xml

anasik commented on test/Microsoft.AspNetCore.OData.E2E.Tests/Lists/ListsTest.cs:
```diff
@@ -0,0 +1,301 @@
+        {
+            string currentDirectory = Environment.CurrentDirectory;
+            services.ConfigureControllers(typeof(ProductsController), typeof(MetadataController));
+            var keepAliveConnection = new SqliteConnection("DataSource=:memory:");
```
I'm really confused and curious about how it was working all this time when I was missing that directive. In fact, I opened my IDE and just before I fixed it I could even see an error on that line.

anasik commented:
> @anasik Also please run all the tests locally and ensure they are passing. This is to ensure this change does not break anything.

Umm, I don't think that's possible. For some reason, since day one, most tests have been failing for me. I think it has something to do with my setup. Perhaps the fact that I am on a mac and unable to install SQL Server.

WanjohiSammy commented:
> Umm, I don't think that's possible. For some reason, since day one, most tests have been failing for me. I think it has something to do with my setup. Perhaps the fact that I am on a mac and unable to install SQL Server.

We have some `Unit Tests` that were not failing and now, with this change, they are failing both locally and, on the pipeline. 

If you checkout to `release-8.x`, run `clean solution`, rebuild and run tests, all the tests should pass unless it is config issue with your environment.

We recommend to first run all tests locally and ensure all of them are passing.

anasik commented:
> If you checkout to `release-8.x`, run `clean solution`, rebuild and run tests, all the tests should pass unless it is config issue with your environment.

That is exactly what I am trying to tell you. There are config issues with my environment that I cannot fix. If you tell me exactly which tests are failing that weren't failing before, then I can help you better.

WanjohiSammy commented:
> That is exactly what I am trying to tell you. There are config issues with my environment that I cannot fix. If you tell me exactly which tests are failing that weren't failing before, then I can help you better.

They are too many. Let me share screenshots and I hope you will be able locate them:

![image](https://github.com/user-attachments/assets/a4c77b63-1ee1-439a-8059-b9756961bff1)

### E2E Tests (net6.0, netcoreapp3.1 etc)
![image](https://github.com/user-attachments/assets/0fe0ca01-fa56-48fc-b954-671a781d399c)

### Unit Tests (net6.0, netcoreapp3.1 etc)
![image](https://github.com/user-attachments/assets/e07a2c6d-882f-463d-8372-c580ed50fc4b)

### E2E Tests (net8.0)
![image](https://github.com/user-attachments/assets/131c01e2-9daa-4b8f-b287-7bb1c03fa69c)

WanjohiSammy commented:
> @WanjohiSammy What I don't understand, is how come there's 402 more tests on my branch than there are on `release-8.x` ?

Not sure about this but you can try to pull the latest changes in the `release-8.x` and compare.

If you’re using a repo forked from `OData/AspNetCoreOData` repo, use the following commands to align your tracking branches with the remote branches.

Where `upstream` is the alias for the OData/AspNetCoreOData repo, and `origin` is the alias for the <YourGitHubAccount>/AspNetCoreOData repo,

```bash
git checkout main
git fetch upstream
git branch -m main release-8.x
git rebase upstream/release-8.x
git push --set-upstream origin release-8.x
```

Then,
```bash
git branch -m dev-9.x main
git checkout main
git rebase upstream/main
git fetch origin
git branch -u origin/main main
git remote set-head origin -a
git push –force
```

anasik commented:
> > @WanjohiSammy What I don't understand, is how come there's 402 more tests on my branch than there are on `release-8.x` ?
> 
> Not sure about this but you can try to pull the latest changes in the `release-8.x` and compare.
> 
> If you’re using a repo forked from `OData/AspNetCoreOData` repo, use the following commands to align your tracking branches with the remote branches.
> 
> Where `upstream` is the alias for the OData/AspNetCoreOData repo, and `origin` is the alias for the /AspNetCoreOData repo,
> 
> ```shell
> git checkout main
> git fetch upstream
> git branch -m main release-8.x
> git rebase upstream/release-8.x
> git push --set-upstream origin release-8.x
> ```
> 
> Then,
> 
> ```shell
> git branch -m dev-9.x main
> git checkout main
> git rebase upstream/main
> git fetch origin
> git branch -u origin/main main
> git remote set-head origin -a
> git push –force
> ```

This seems like trouble. Led to a lot of conflicts. Should I just accept all the upstream changes for every conflict?

anasik commented:
@WanjohiSammy btw, I undid my change in the `SelectExpandBinder` and ran the tests again. Apparently a lot of the tests are failing because of NET8 and not because of my change. I know this applies to the entire `Cast` tests at least. 

For all 30 of them, this line is failing:
`            HttpResponseMessage response = await Client.GetAsync(requestUri);
`
Microsoft.AspNetCore.OData.E2E.Tests/Cast/CastTest.cs:line 108

and the error is: 
`The format of value '' is invalid.`

anasik commented:
@WanjohiSammy, please run the tests again. This time, a lot of the previously failing tests would pass. I need to know which ones remain.

WanjohiSammy commented:
> @WanjohiSammy, please run the tests again. This time, a lot of the previously failing tests would pass. I need to know which ones remain.
> 
> Furthermore, me being on an m1 mac means it's very difficult for me to run the 3.1 tests so I am gonna need a little help with debugging those.
> 
> And I have absolutely no idea how to run the LocalDB tests.

The failing tests have reduced to `49`.

anasik commented:
> > @WanjohiSammy, please run the tests again. This time, a lot of the previously failing tests would pass. I need to know which ones remain.
> > Furthermore, me being on an m1 mac means it's very difficult for me to run the 3.1 tests so I am gonna need a little help with debugging those.
> > And I have absolutely no idea how to run the LocalDB tests.
> 
> The failing tests have reduced to `49`.

Thank you! Now can you perhaps maybe give me as much detail as possible about which ones are failing and why? I wish I could recreate this myself by I'm severely limited by my setup. Perhaps you can export the test results?

WanjohiSammy commented:
> > The failing tests have reduced to `49`.
> 
> Thank you! Now can you perhaps maybe give me as much detail as possible about which ones are failing and why? I wish I could recreate this myself by I'm severely limited by my setup. Perhaps you can export the test results?

To provide a few:

```bash
 Microsoft.AspNetCore.OData.E2E.Tests.NavigationPropertyOnComplexType.SelectImprovementOnComplexTypeTests.QueryEntityWithSelectOnCollectionPrimitivePropertyOfComplexPropertyWithNestedTopAndSkip(select: "HomeLocation/Emails($top=1;$skip=1)", value: "\"Emails\":[\"E6\"]")
   Source: SelectImprovementOnComplexTypeTests.cs line 319

  Message: 
Assert.Equal() Failure
                                 ↓ (pos 113)
Expected: ···ation":{"Emails":["E6"]}}
Actual:   ···ation":{"Emails":["E9","E6","E8"]}}
                                 ↑ (pos 113)

 Microsoft.AspNetCore.OData.E2E.Tests.NavigationPropertyOnComplexType.SelectImprovementOnComplexTypeTests.QueryEntityWithSelectOnCollectionPrimitivePropertyOfComplexPropertyWithNestedTopAndSkip(select: "HomeLocation/Emails($top=2)", value: "\"Emails\":[\"E9\",\"E6\"]")
   Source: SelectImprovementOnComplexTypeTests.cs line 319

  Message: 
Assert.Equal() Failure
                                 ↓ (pos 120)
Expected: ···{"Emails":["E9","E6"]}}
Actual:   ···{"Emails":["E9","E6","E8"]}}
                                 ↑ (pos 120)

 Microsoft.AspNetCore.OData.E2E.Tests.NavigationPropertyOnComplexType.SelectImprovementOnComplexTypeTests.QueryEntityWithSelectOnCollectionPrimitivePropertyWithNestedTopAndSkip(select: "Taxes($top=1;$skip=1)", value: "\"Taxes\":[5]")
   Source: SelectImprovementOnComplexTypeTests.cs line 266

  Message: 
Assert.Equal() Failure
                                 ↓ (pos 80)
Expected: ···)/$entity","Taxes":[5]}
Actual:   ···)/$entity","Taxes":[7,5,9]}
                                 ↑ (pos 80)

 Microsoft.AspNetCore.OData.E2E.Tests.NavigationPropertyOnComplexType.SelectImprovementOnComplexTypeTests.QueryEntityWithSelectOnCollectionPrimitivePropertyWithNestedTopAndSkip(select: "Taxes($top=2)", value: "\"Taxes\":[7,5]")
   Source: SelectImprovementOnComplexTypeTests.cs line 266

  Message: 
Assert.Equal() Failure
                                 ↓ (pos 83)
Expected: ···entity","Taxes":[7,5]}
Actual:   ···entity","Taxes":[7,5,9]}
                                 ↑ (pos 83)

 Microsoft.AspNetCore.OData.E2E.Tests.ODataOrderByTest.OrderByAdvancedTest.DollarOrderBy_UsingDollarThisWithinDollarSelect_OnCollection
   Source: OrderByAdvancedTest.cs line 204

  Message: 
Assert.Equal() Failure
                     ↓ (pos 11)
Expected: {"Grades":[1,2,7,8,9]}
Actual:   {"Grades":[9,8,7,1,2]}
                     ↑ (pos 11)


 Microsoft.AspNetCore.OData.E2E.Tests.Routing.ODataRoutingTests.RequestContainsValidValues_CanRouteToCorrectEndpoint(key: "other", isODataRouting: False, expect: "{\"id\":11,\"name\":\"Non-OData Product\"}")
   Source: ODataRoutingTests.cs line 45

  Message: 
Assert.Equal() Failure
            ↓ (pos 2)
Expected: {"id":11,"name":"Non-OData Product"}
Actual:   {"name":"Non-OData Product","id":11}
            ↑ (pos 2)


 Microsoft.AspNetCore.OData.E2E.Tests.ServerSidePaging.ServerSidePagingTests.VerifyClientSidePagingAppliedToNonEntityCollections
   Source: ServerSidePagingTests.cs line 480

  Message: 
The collection contained 3 matching element(s) instead of 1.


 Microsoft.AspNetCore.OData.Tests.Query.Expressions.SelectExpandBinderTest.ProjectAsWrapper_Element_ProjectedValueContains_SelectedCollectStructuralProperties(select: "Emails($orderby=$this desc)", expect: ["E3", "E2", "E1"])
   Source: SelectExpandBinderTest.cs line 654

  Message: 
Assert.Equal() Failure
Expected: String[] ["E3", "E2", "E1"]
Actual:   String[] ["E1", "E3", "E2"]


 Microsoft.AspNetCore.OData.Tests.Query.Expressions.SelectExpandBinderTest.ProjectAsWrapper_Element_ProjectedValueContains_SelectedCollectStructuralProperties(select: "Emails($orderby=$this)", expect: ["E1", "E2", "E3"])
   Source: SelectExpandBinderTest.cs line 654

  Message: 
Assert.Equal() Failure
Expected: String[] ["E1", "E2", "E3"]
Actual:   String[] ["E1", "E3", "E2"]


 Microsoft.AspNetCore.OData.Tests.Query.Expressions.SelectExpandBinderTest.ProjectAsWrapper_Element_ProjectedValueContains_SelectedSubCollectionStructuralProperties(select: "HomeAddress($select=Codes($orderby=$this desc))", expect: "C4,C2,C1")
   Source: SelectExpandBinderTest.cs line 789

  Message: 
Assert.Equal() Failure
           ↓ (pos 1)
Expected: C4,C2,C1
Actual:   C1,C4,C2
           ↑ (pos 1)


 Microsoft.AspNetCore.OData.Tests.Query.Expressions.SelectExpandBinderTest.ProjectAsWrapper_Element_ProjectedValueContains_SelectedSubCollectionStructuralProperties(select: "HomeAddress($select=Codes($top=1;$skip=2))", expect: "C2")
   Source: SelectExpandBinderTest.cs line 789

  Message: 
Assert.Equal() Failure
           ↓ (pos 1)
Expected: C2
Actual:   C1,C4,C2
           ↑ (pos 1)
```

anasik commented:
@WanjohiSammy, Now I think we should be good.

WanjohiSammy commented on test/Microsoft.AspNetCore.OData.E2E.Tests/Microsoft.AspNetCore.OData.E2E.Tests.csproj:
```diff
@@ -27,21 +26,33 @@
   </ItemGroup>
-    
+  <ItemGroup Condition="'$(TargetFramework)' == 'net8.0' ">
+    <PackageReference Include="Microsoft.AspNetCore.Mvc.Testing" Version="8.0.3" />
```
Why using `8.0.3` instead of `8.0.8`
```suggestion
    <PackageReference Include="Microsoft.AspNetCore.Mvc.Testing" Version="8.0.8" />
    <PackageReference Include="Microsoft.EntityFrameworkCore" Version="8.0.8" />
    <PackageReference Include="Microsoft.EntityFrameworkCore.InMemory" Version="8.0.8" />
    <PackageReference Include="Microsoft.EntityFrameworkCore.Sqlite" Version="8.0.8" />
    <PackageReference Include="Microsoft.EntityFrameworkCore.SqlServer" Version="8.0.8" />
    <PackageReference Include="Microsoft.EntityFrameworkCore.Proxies" Version="8.0.8" />
```

WanjohiSammy commented on test/Microsoft.AspNetCore.OData.E2E.Tests/Microsoft.AspNetCore.OData.E2E.Tests.csproj:
```diff
@@ -27,21 +26,38 @@
+    <PackageReference Include="Microsoft.EntityFrameworkCore.Sqlite" Version="8.0.8" />
+    <PackageReference Include="Microsoft.EntityFrameworkCore.SqlServer" Version="8.0.8" />
+    <PackageReference Include="Microsoft.EntityFrameworkCore.Proxies" Version="8.0.8" />
+    <PackageReference Include="Microsoft.EntityFrameworkCore" Version="8.0.3" />
```
Remove the 8.0.3. You already have the 8.0.8 version.
```suggestion
```
In my previous comment, wanted you to use 8.0.8 version instead of 8.0.3

WanjohiSammy commented on test/Microsoft.AspNetCore.OData.E2E.Tests/Microsoft.AspNetCore.OData.E2E.Tests.csproj:
```diff
@@ -27,21 +26,38 @@
+    <PackageReference Include="Microsoft.EntityFrameworkCore.SqlServer" Version="8.0.8" />
+    <PackageReference Include="Microsoft.EntityFrameworkCore.Proxies" Version="8.0.8" />
+    <PackageReference Include="Microsoft.EntityFrameworkCore" Version="8.0.3" />
+    <PackageReference Include="Microsoft.EntityFrameworkCore.InMemory" Version="8.0.3" />
```
Remove this as well
```suggestion
```

WanjohiSammy commented on test/Microsoft.AspNetCore.OData.E2E.Tests/Microsoft.AspNetCore.OData.E2E.Tests.csproj:
```diff
@@ -27,21 +26,38 @@
+    <PackageReference Include="Microsoft.EntityFrameworkCore.Proxies" Version="8.0.8" />
+    <PackageReference Include="Microsoft.EntityFrameworkCore" Version="8.0.3" />
+    <PackageReference Include="Microsoft.EntityFrameworkCore.InMemory" Version="8.0.3" />
+    <PackageReference Include="Microsoft.EntityFrameworkCore.Sqlite" Version="8.0.3" />
```
Remove this
```suggestion
```

anasik replied:
We're using it I think.

WanjohiSammy replied:
You can remove `<PackageReference Include="Microsoft.EntityFrameworkCore.Sqlite" Version="8.0.3" />` and keep `<PackageReference Include="Microsoft.EntityFrameworkCore.Sqlite" Version="8.0.8" />`. I think you do not need both versions of the same dependency.

WanjohiSammy commented on test/Microsoft.AspNetCore.OData.E2E.Tests/Microsoft.AspNetCore.OData.E2E.Tests.csproj:
```diff
@@ -27,21 +26,38 @@
+    <PackageReference Include="Microsoft.EntityFrameworkCore" Version="8.0.3" />
+    <PackageReference Include="Microsoft.EntityFrameworkCore.InMemory" Version="8.0.3" />
+    <PackageReference Include="Microsoft.EntityFrameworkCore.Sqlite" Version="8.0.3" />
+    <PackageReference Include="Microsoft.EntityFrameworkCore.SqlServer" Version="8.0.3" />
```
Remove this since you already have 8.0.8
```suggestion
```

WanjohiSammy commented on test/Microsoft.AspNetCore.OData.E2E.Tests/Microsoft.AspNetCore.OData.E2E.Tests.csproj:
```diff
@@ -27,21 +26,38 @@
+    <PackageReference Include="Microsoft.EntityFrameworkCore.InMemory" Version="8.0.3" />
+    <PackageReference Include="Microsoft.EntityFrameworkCore.Sqlite" Version="8.0.3" />
+    <PackageReference Include="Microsoft.EntityFrameworkCore.SqlServer" Version="8.0.3" />
+    <PackageReference Include="Microsoft.EntityFrameworkCore.Proxies" Version="8.0.3" />
```
Remove this.
```suggestion
```

WanjohiSammy commented on test/Microsoft.AspNetCore.OData.E2E.Tests/Microsoft.AspNetCore.OData.E2E.Tests.csproj:
```diff
@@ -27,21 +26,33 @@
-    <ProjectReference Include="..\..\src\Microsoft.AspNetCore.OData\Microsoft.AspNetCore.OData.csproj" />
-    <ProjectReference Include="..\Microsoft.AspNetCore.OData.TestCommon\Microsoft.AspNetCore.OData.TestCommon.csproj" />
+    <ProjectReference
+      Include="..\..\src\Microsoft.AspNetCore.OData\Microsoft.AspNetCore.OData.csproj" />
```
Return this to one line since there is no change here
```suggestion
      <ProjectReference Include="..\..\src\Microsoft.AspNetCore.OData\Microsoft.AspNetCore.OData.csproj" />
```

WanjohiSammy commented on test/Microsoft.AspNetCore.OData.E2E.Tests/Microsoft.AspNetCore.OData.E2E.Tests.csproj:
```diff
@@ -27,21 +26,33 @@
+    <ProjectReference
+      Include="..\..\src\Microsoft.AspNetCore.OData\Microsoft.AspNetCore.OData.csproj" />
+    <ProjectReference
+      Include="..\Microsoft.AspNetCore.OData.TestCommon\Microsoft.AspNetCore.OData.TestCommon.csproj" />
```
Same here. What is the change here?

anasik replied:
No idea

WanjohiSammy replied:
You can undo the change here

anasik commented:
@WanjohiSammy, is there an ETA on this?

WanjohiSammy commented on src/Microsoft.AspNetCore.OData/Common/TypeHelper.cs:
```diff
@@ -252,6 +252,26 @@ Type collectionInterface
+        {
+            return type.IsPrimitive 
+                   || type == typeof(string)
+                   || type == typeof(Uri)
```
Team had a concern about `Uri`
We currently do not support Uri as a primitive type  @gathogojr

anasik replied:
Then why didn't you just say so?  thought we'd already discussed this. But lets discuss it again. 

Personally, I don't consider even `string` a primitive type, and neither does anyone else in our profession. But for the purpose of the task at hand, it is and the documentation for ASP NET Core says so. Just like it says the same thing for `Uri`. If it's too concerning, we can remove this for now and wait for someone else to attempt to add it back a few PRs down the line. 

With all due respect sir, this is actually a **pretty major bug**, on our last stable version that has been **fixed since June** but we are not rolling it out for unknown reasons. Me and my team are personally suffering from this enough to be using a fork in our project. And there could be people out there who have been experiencing this ever since it was first released. 

All I am saying is I would appreciate us doing anything we can to **move things along**. Which means that if a line of code that adds an extra case for the URL type bothers the team so much, **I don't mind removing it** because that way the 99 other types don't retain the bug for another 6 months.

WanjohiSammy replied:
> @WanjohiSammy Then why didn't you just say so? thought we'd already discussed this. But lets discuss it again.
> 
> Personally, I don't consider even `string` a primitive type, and neither does anyone else in our profession. But for the purpose of the task at hand, it is and the documentation for ASP NET Core says so. Just like it says the same thing for `Uri`. If it's too concerning, we can remove this for now and wait for someone else to attempt to add it back a few PRs down the line.
> 
> With all due respect sir, this is actually a **pretty major bug**, on our last stable version, that has been **fixed since June** but we are not rolling it out for unknown reasons. Me and my team are personally suffering from this enough to be using a fork in our project. And there could be people out there who have been experiencing this ever since it was first released.
> 
> All I am saying is I would appreciate us doing anything we can to **move things along**. Which means that if a line of code that adds an extra case for the URL type bothers the team so much, **I don't mind removing it** because that way the 99 other types don't retain the bug for another 6 months.

This looks good to me.
@gathogojr, @habbes  what do you think about this change. 
@anasik apologies

WanjohiSammy commented on src/Microsoft.AspNetCore.OData/Common/TypeHelper.cs:
```diff
@@ -252,6 +252,26 @@ Type collectionInterface
+        {
+            return type.IsPrimitive 
+                   || type == typeof(string)
+                   || type == typeof(Uri)
```
Team had a concern about `Uri`
We currently do not support Uri as a primitive type  @gathogojr
